torch/distributed/_functional_collectives_impl.py

```diff
@@ -60,34 +75,36 @@ def _register_tensor_work(tensor_or_list, work_or_list):
     if not isinstance(work_or_list, list):
         reg = _WaitRegistration(work_or_list)
         for tensor in tensor_or_list:
-            reg._register(tensor)
+            reg._register_tensor_ptr(tensor.data_ptr())
     else:
         for tensor, work in zip(tensor_or_list, work_or_list):
             reg = _WaitRegistration(work)
-            reg._register(tensor)
+            reg._register_tensor_ptr(tensor.data_ptr())
+
+
 
 def _wait_reg_dec(ptr, wait_reg):
     wait_reg.decrement_live_tensor(ptr)
 
-def _register_wait_tensor(tensor):
+def _register_tensor_wrapper(tensor) -> None:
     global data_ptr_to_work
+    data_ptr = tensor.elem.data_ptr()
     # Note: we should NEVER try to trace this, bc it registers runtime stuff during trace.
     # Instead, backends must call this themselves when implementing traced collectives.
-    wait_reg = data_ptr_to_work.get(tensor.data_ptr(), None)
+    wait_reg = data_ptr_to_work.get(data_ptr, None)
     if wait_reg is None:
         warnings.warn(
-            "Trying to register finalizers to AsyncCollectiveTensor but the inner tensor is already gone"
+            "Trying to register finalizer to AsyncCollectiveTensor but the inner tensor is already gone"
         )
     else:
         # We force the collective to be waited in the case this tensor goes away to reduce the change of deadlocks.
-        # NOTE: we register the callback to the inner tensor of ACT, instead of the ACT wrapper
-        # class, for two reasons:
-        # 1. sometimes ACT need to continue propagation by unwrapping, then rewrapping with a new
-        #    ACT instance, if we register the callback to the ACT wrapper class, the callback will
-        #    be called when the ACT wrapper class is garbage collected, which is not what we want.
-        # 2. register callback to the inner tensor of ACT could achieve the same purpose as register
-        #    on the ACT wrapper, as it's the inner tensor who is going to be GCed need to wait.
-        weakref.finalize(tensor, _wait_reg_dec, tensor.data_ptr(), wait_reg)
+        # NOTE: we register the callback to the ACT wrapper class, for the following reasons:
+        # 1. The inner tensor is referenced by the associated Work object, so it's uncollective until we release the
+        #  associated work object
+        # 2. There's a n-to-1 relationship between wrappers and inner tensor due to non-waitable ops like view()
+        wait_reg._record_wrapper(data_ptr)
+        weakref.finalize(tensor, _wait_reg_dec, data_ptr, wait_reg)
```

Comment:
One thing I don't understand is that we already call `self.cleanup()` in `wait()` but why do we still register a callback here. Would this be duplicate?